### PR TITLE
Additionnal features for the character controller template.

### PR DESCRIPTION
**Pull request by Gabriel-Dalmon**
- Loading screen
- Pointerlock management
- Sensitivity settings
- Gamepad detection and sensitivity adjustment
- Control keys edition
- Spawn position control
- Inactivity Popup
- Disconnection popup on onConnectionClose & reconnection buttons

### CODE REVIEW

File: scripts/settings.js [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    "\x14": "CAPSLOCK",
+    "&": "ARROW UP",
+    "(": "ARROW DOWN",
+    "%": "ARROW LEFT",
+    "'": "ARROW RIGHT",
+}
+
+//------------------------------------------------------------------------------
+export async function InitDeviceDetection() {
+    if(navigator.getGamepads().length > 0) {
```
Comment (nathalie-raffray) [inline scripts/settings.js:29]: when i tried this, i had disconnected the gamepad from my laptop, but navigator.getGamepads() somehow returned an array of length four : [null, null, null, null] 🙃

Comment (nathalie-raffray) [reply scripts/settings.js:29]: apparently navigator.getGamepads() can return null elements, so you should check whether there are non null elements here

File: scripts/settings.js [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    const device = GetDevice();
+    const sensitivitySetting = GetSensitivity();
+    
+    // We adjust the sensitivity depending on the device. The joysticks on 
+    // gamepads produce less sensitive values than the mouse. The new 
+    // sensitivity for gamepad has to be higher. Around ~0.x sensitivity for 
+    // mouse and ~x.0 for gamepad within the current asset script for character 
+    // controller camera management.
+    if ( device ==  "gamepad" ){
+        var newSensitivity = sensitivitySetting / 5;
```
Comment (nathalie-raffray) [inline scripts/settings.js:75]: most web devs dont use var because of its idiosyncratic interpretation of scope lol

Comment (nathalie-raffray) [reply scripts/settings.js:75]: best practice is to use const wherever possible, and if not use let

Comment (nathalie-raffray) [reply scripts/settings.js:75]: in this case you can use let:
```
let newSensitivity;
if(device === 'gamepad') {
    newSensitivity = ...
```

File: scripts/settings.js [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    // We recover the device and the sensitivity from the settings modal inputs.
+    const device = GetDevice();
+    const sensitivitySetting = GetSensitivity();
+    
+    // We adjust the sensitivity depending on the device. The joysticks on 
+    // gamepads produce less sensitive values than the mouse. The new 
+    // sensitivity for gamepad has to be higher. Around ~0.x sensitivity for 
+    // mouse and ~x.0 for gamepad within the current asset script for character 
+    // controller camera management.
+    if ( device ==  "gamepad" ){
```
Comment (nathalie-raffray) [inline scripts/settings.js:74]: best practice in js is to always use === instead of ==

File: scripts/settings.js [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    clientData.characterController.setScriptInputValues(
+        characterControllerScriptUUID, 
+        { sensitivity: newSensitivity }
+    );
+
+    // This should be removed when setScriptInputValues() is fixed, it's just 
+    // that since setScriptInputValues doesn't return a promise, we must make 
+    // sure that the script inputs have been updated before reassignign client 
+    // to script.
+    await SDK3DVerse.engineAPI.findEntitiesByEUID(characterControllerSceneUUID);
```
Comment (nathalie-raffray) [inline scripts/settings.js:92]: it doesnt work if you remove this line?

File: config.js [hunk 5]
```diff
@@ -1,3 +1,7 @@
 export const publicToken = %YOUR_PUBLIC_TOKEN%;
 export const mainSceneUUID = %YOUR_MAIN_SCENE_UUID%;
 export const characterControllerSceneUUID = %YOUR_CHARACTER_CONTROLLER_SCENE_UUID%;
+export const characterControllerScriptUUID = %YOUR_CHARACTER_CONTROLLER_SCRIPT_UUID%;
+export const spawnPosition = [0,2,0];
+
+export let clientData = { characterController: null, };
```
Comment (nathalie-raffray) [inline config.js:7]: This shouldnt be in the config, the config is where a user can input some static data that changes the inputs of the app

File: scripts/settings.js [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+//------------------------------------------------------------------------------
+export function GetSensitivity() {
+    return document.getElementById("sensitivity-slider").value;
+}
+
+//------------------------------------------------------------------------------
+export function GetDevice() {
+    return document.getElementById("device").value;
```
Comment (nathalie-raffray) [inline scripts/settings.js:106]: device doesnt need to be a hidden DOM element, it can just be a global variable

File: config.js [hunk 7]
```diff
@@ -1,3 +1,7 @@
 export const publicToken = %YOUR_PUBLIC_TOKEN%;
 export const mainSceneUUID = %YOUR_MAIN_SCENE_UUID%;
 export const characterControllerSceneUUID = %YOUR_CHARACTER_CONTROLLER_SCENE_UUID%;
+export const characterControllerScriptUUID = %YOUR_CHARACTER_CONTROLLER_SCRIPT_UUID%;
```
Comment (nathalie-raffray) [inline config.js:4]: remove this, and get the script uuid from the script map component
 its better to keep config short and sweet, cus its annoying for a user who doesnt have much experience in 3dverse to go look for assets (and i think the script isnt in public folder like character controller scene and main scene)

File: scripts/settings.js (hunk 4 above)
Comment (Gabriel-Dalmon) [reply scripts/settings.js:92]: I read a bit of the SDK source today and I think I need to use something similar to this instead : 

`SDK3DVerse.notifier.on('onEntitiesUpdated', assignClientToCharacterController);

function assignClientToCharacterController() {
    SDK3DVerse.engineAPI.assignClientToScripts(clientData.characterController));
    SDK3DVerse.notifier.off('onEntitiesUpdated', assignClientToCharacterController);
}`

Basically, I need to make sure `setScriptInputValues()` effect has occured before calling the assignClientToScript.

File: config.js (hunk 5 above)
Comment (Gabriel-Dalmon) [reply config.js:7]: Yes, I see, it makes sense. Should I make characterController a global variable similarly to the device or is it preferable to pass as argument to each functions that needs it ? 

I'd started with passing it as argument initially but it felt like the more parameters the functions have the harder it makes it to manage the event listeners, even though nothing impossible.

File: scripts/settings.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    } else { //if ( device === "mouse" ){
+        newSensitivity = sensitivitySetting / 100;
+    }
+
+    // By default, the camera orientation is reset when calling 
+    // "setScriptInputValues()".  We pass the current pitch and yaw of the 
+    // camera to the Asset Script to keep the camera orientation.
+    // Camera orientation can be obtained from the viewport orientation.
+    const activeViewports = SDK3DVerse.engineAPI.cameraAPI.getActiveViewports();
+    const viewportOrientation = activeViewports[0].getTransform().orientation;
```
Comment (nathalie-raffray) [inline scripts/settings.js:84]: fyi this function is called getViewports now ! (btw i fixed this bug, but its still in dev, will ping you once its in prod)

File: scripts/settings.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    return document.getElementById("sensitivity-slider").value;
+}
+
+//------------------------------------------------------------------------------
+export function InitDeviceDetection() {
+    ResetGamepadDetection();
+}
+
+//------------------------------------------------------------------------------
+async function ResetGamepadDetection() {
```
Comment (nathalie-raffray) [inline scripts/settings.js:122]: this doesnt need to be async, window.addEventListener isnt async, it just registers an async function to 'gamepadconnected'

you can fix for the other functions like this one

**Review by nathalie-raffray — COMMENTED** [review]
just a few minor things here and there. gg

File: scripts/main.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    SDK3DVerse.setInactivityCallback(() => {
+        setTimeout(() => { 
+            ShowDisconnectedPopup() 
+        }, 30 * 1000); 
+    });
+
+    // The following does the same but in case the disconnection is 
+    // requested by the server.
+    SDK3DVerse.notifier.on("onConnectionClosed", () => { 
+        SDK3DVerse.setInactivityCallback(()=>{});
```
Comment (nathalie-raffray) [inline scripts/main.js:134]: to see with @apetitescure (alex) if this is necessary here

Comment (Gabriel-Dalmon) [reply scripts/main.js:134]: I think `setInactivityCallback()` is useless when using onConnectionClosed event because the event will be triggered anyway. 

Instead of using the `setInactivityCallback()` for a disconnected popup that shows up after the 30 seconds timeout, I can make the popup be specific for inactivity state that directly appears after the 5 minutes of inactivity before updating it to disconnected state when onConnectionClosed is called.

File: scripts/settings.js (hunk 8 above)
Comment (Gabriel-Dalmon) [reply scripts/settings.js:84]: getViewports doesn't seem to be in prod either. I can only use getActiveViewports when importing the SDK with `<script src="https://cdn.3dverse.com/legacy/sdk/stable/SDK3DVerse.js"></script>`

File: scripts/main.js [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+import { 
+  InitDeviceDetection, 
+  InitControlKeySettings,
+  AdjustDeviceSensitivity, 
+  OpenSettingsModal
+} from "./settings.js";
+
+//------------------------------------------------------------------------------
+window.characterController = "HEY";
```
[the remark this answers is not in the document]
Comment (Gabriel-Dalmon) [reply scripts/main.js:19]: Should I use a specific globals.js file in which I declare my globals (characterController & device) ? Or place the characterController definition in settings.js even though the variable is set in main.js.

Otherwise I believe I can't declare globals from the main.js and import them in other files without storing it in another already global object, such as the window. I guess it's related to circular imports.

That's why it initially was in the config.js.

File: scripts/settings.js [hunk 12]
```diff
@@ -163,3 +161,8 @@ export async function CloseSettingsModal() {
     const close = document.getElementById("close");
     close.removeEventListener('click', CloseSettingsModal); 
 }
+
+//------------------------------------------------------------------------------
+export function SetCharacterController(value) {
+    characterController = value;
+}
```
Comment (Gabriel-Dalmon) [inline scripts/settings.js:168]: This setter is used in main.js at application initialization to set the global variable storing the character controller.

I'm not sure wether this is good or not, to define characterController global variable directly in the `settings.js` file even though `settings.js` only reads/calls character controller methods and `main.js` inits/writes the character controller into the global variable.

I believe I can't define the global variable in main.js and import it in settings.js due to circular imports/`main.js` already importing from `settings.js`.

File: scripts/main.js (hunk 10 above)
Comment (apetitescure) [reply scripts/main.js:135]: Effectivement c'etait bien 30s, et non 60s, je m'etais trompé.
Seulement, le concept de ce callback, c'est de permettre d'offrir une derniere chance a l'utilisateur de notifier de sa présence, pas de gérer la déconnexion elle meme.

Peu importe la raison de la connexion, `onConnectionClosed` sera toujours appelé, que ce soit pour un timeout ou une autre raison.

TL;DR tu peux juste completement enlever ce callback, et juste gérer le `onConnectionClosed`

File: scripts/main.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+//------------------------------------------------------------------------------
+function initPointerLockEvents(characterController) {
+    const canvas = document.getElementById("display-canvas");
+    canvas.addEventListener('mousedown', lockPointer);
+
+    // If the user leaves the pointerlock, we open the settings popup and
+    // disable their influence over the character controller.
+    document.addEventListener('keydown', (event) => {
+        if(event.code === 'Escape') {
```
Comment (nathalie-raffray) [inline scripts/main.js:154]: can you also make it possible to close the settings modal by pressing escape?

**Review by nathalie-raffray — COMMENTED** [review]
new loading screen looks nice! (we might have to remove it tho once bens loading screen gets to stable). everythings very simple/clean 👍

File: scripts/main.js [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+      child.isAttached("camera")
+    );
+
+    // We need to assign the current client to the first person controller
+    // script which is attached to the firstPersonController entity.
+    // This allows the script to know which client inputs it should read.
+    SDK3DVerse.engineAPI.assignClientToScripts(firstPersonController);
+
+    // Finally set the first person camera as the main camera.
+    await SDK3DVerse.setMainCamera(firstPersonCamera);
```
Comment (nathalie-raffray) [inline scripts/main.js:91]: setMainCamera is in cameraAPI now

File: scripts/settings.js [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        for (const key of PHYSICAL_ACTION_KEYS[action]) {
+            displayedKeys += (await getActionKey(key)) + " + ";
+        }
+        displayedKeys = displayedKeys.slice(0, -3);
+        element.innerHTML = displayedKeys;
+    });
+}
+
+//------------------------------------------------------------------------------
+export async function adjustDeviceSensitivity(characterController) {
```
Comment (nathalie-raffray) [inline scripts/settings.js:45]: remove async

File: scripts/settings.js [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+//------------------------------------------------------------------------------
+// We iterate over the action keys elements and set the DOM Elements from 
+// the settings modal to the corresponding keys based on the keyboard layout
+// map if available.
+export function initControlKeySettings() {
+    const actionKeysElements = document.getElementsByClassName("action-keys");
+    let action, displayedKeys;
+    Array.from(actionKeysElements).forEach(async (element) => {
+        action = element.getAttribute("data-action")
```
Comment (nathalie-raffray) [inline scripts/settings.js:34]: ;
